src/main/java/org/jvnet/hudson/reactor/Reactor.java

```diff
@@ -205,13 +209,17 @@ public synchronized void addAll(Iterable<? extends Task> _tasks) {
         for (final Task t : _tasks) {
             Node n = new Node(new Runnable() {
                 public void run() {
-                    listener.onTaskStarted(t);
                     try {
+                        listener.onTaskStarted(t);
                         runTask(t);
                         listener.onTaskCompleted(t);
                     } catch (Throwable x) {
                         boolean fatal = t.failureIsFatal();
-                        listener.onTaskFailed(t,x, fatal);
+                        try {
+                            listener.onTaskFailed(t, x, fatal);
+                        } catch(Throwable x2) {
```

Comment:
I would suggest throwing a new `TunnelException` with keeping the original error by using `addSuppressedException()`

Reply:
was thinking about providing both exceptions, but did not know about suppressed Exceptions. will do the change.